web/src/bot_data.ts

```diff
@@ -0,0 +1,133 @@
+    base_url: string;
+    interface: number;
+    config_data: {key: string};
+    service_name: string;
+    token: string;
+};
+
+export type CleanBot = {
+    api_key: string;
+    avatar_url: string;
+    bot_type: number;
+    default_all_public_streams: boolean;
+    default_events_register_stream: string | undefined;
+    default_sending_stream: string | undefined;
+    email: string;
+    full_name: string;
+    is_active: boolean;
+    owner: string;
+    owner_id: number;
+    user_id: number;
+};
+
+export type Bot = CleanBot & {
+    services: Service[];
+    extra?: string;
+};
+
+export type BotDataParams = {
+    realm_bots: Bot[];
+};
+
+const bots = new Map<number, CleanBot>();
+
+const bot_fields = [
+    "api_key",
+    "avatar_url",
+    "bot_type",
+    "default_all_public_streams",
+    "default_events_register_stream",
+    "default_sending_stream",
+    "email",
+    "full_name",
+    "is_active",
+    "owner", // TODO: eliminate
+    "owner_id",
+    "user_id",
+];
+
+const services = new Map<number, Service[]>();
+const services_fields = ["base_url", "interface", "config_data", "service_name", "token"];
+
+export function all_user_ids(): number[] {
+    return [...bots.keys()];
+}
+
+export function add(bot: Bot): void {
+    const clean_bot = _.pick(bot, bot_fields) as CleanBot;
+    bots.set(bot.user_id, clean_bot);
+    const clean_services = bot.services.map(
+        (service) => _.pick(service, services_fields) as Service,
+    );
+    services.set(bot.user_id, clean_services);
+}
+
+export function deactivate(bot_id: number): void {
+    bots.get(bot_id)!.is_active = false;
+}
+
+export function del(bot_id: number): void {
+    bots.delete(bot_id);
+    services.delete(bot_id);
+}
+
+export function update(bot_id: number, bot_update: Bot): void {
+    const bot = bots.get(bot_id)!;
+    Object.assign(bot, _.pick(bot_update, bot_fields));
+
+    // We currently only support one service per bot.
+    const service = services.get(bot_id)![0];
+    if (bot_update.services !== undefined && bot_update.services.length > 0) {
+        Object.assign(service, _.pick(bot_update.services[0], services_fields));
+    }
+}
+
+export function get_all_bots_for_current_user(): CleanBot[] {
+    const ret = [];
+    for (const bot of bots.values()) {
+        if (people.is_my_user_id(bot.owner_id)) {
+            ret.push(bot);
+        }
+    }
+    return ret;
+}
+
+export function get_editable(): CleanBot[] {
+    const ret = [];
+    for (const bot of bots.values()) {
+        if (bot.is_active && people.is_my_user_id(bot.owner_id)) {
+            ret.push(bot);
+        }
+    }
+    return ret;
+}
+
+export function get_all_bots_owned_by_user(user_id: number): CleanBot[] {
+    const ret = [];
+    for (const bot of bots.values()) {
+        if (bot.owner_id === user_id && bot.is_active) {
+            ret.push(bot);
+        }
+    }
+    return ret;
+}
+
+export function get(bot_id: number): Partial<Bot> | undefined {
```

Comment:
Maybe we can use `CleanBot` here when we have the schema.

Reply:
I agree. I have integrated these keys with zod. Please check.